### DISCUSSION TITLE
Allow `IntPtr` and `UIntPtr` in attributes

### DISCUSSION THREAD

**Discussion by jkrejcha** [dotnet/csharplang#8155] [Language Ideas] [answered]
Currently using `[Foo(0x0)]` as an attribute fails with CS0181
 ```cs
public class FooAttribute
{
    public FooAttribute(IntPtr num)
    {
        // ...
     }
}
```

In addition, [§22.2.4](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes#2224-attribute-parameter-types) only constrains types allowed in attributes to...

> One of the following types: bool, byte, char, double, float, int, long, sbyte, short, string, uint, ulong, ushort.
> * The type object.
> * The type System.Type.
> * Enum types.
> * Single-dimensional arrays of the above types.

and my proposal would allow the usage of `nint` and `nuint` in attributes.

**Comment by HaloFour** [comment, marked as answer]
I don't think it makes sense to use `IntPtr` in attributes as the actual data embedded in the assembly metadata needs to be a known predictable size at compile time.  It can't change to accommodate a different size based on the native word size.  So at best it would still be a `long` or `ulong`, which you could then cast back to an `IntPtr` in code anyway.

**Reply by jkrejcha** [reply]
Ah right I see, this doesn't work for AnyCPU, etc.

**Reply by HaloFour** [reply]
Yeah, it'd be a little weird.  You'd have to pick a fixed size to store the number in the assembly metadata (or have a flag to indicate the number of bits), but the only safe value you could store is 32-bit.  It's not impossible, but it would probably be a bit awkward.

**Reply by tannergooding** [reply]
It's also worth noting that this is a runtime limitation:
> CLS Rule 34: The CLS only allows a subset of the encodings of custom attributes. The only types that shall appear in these encodings are (see Partition IV): System.Type, System.String, System.Char, System.Boolean, System.Byte, System.Int16, System.Int32, System.Int64, System.Single, System.Double, and any enumeration type based on a CLS-compliant base integer type.

There is an interesting bit in that the runtime does allow `nint` based enums and therefore technically `nint` is allowed (albeit indirectly) in an attribute. However, much like `nint` constants, they are restricted to 32-bits and are sign-extended by the runtime up to the actual target size at runtime.

**Reply by HaloFour** [reply]
> There is an interesting bit in that the runtime does allow `nint` based enums and therefore technically `nint` is allowed (albeit indirectly) in an attribute. 

That is interesting, and it doesn't seem like C# supports that today.
